File: nikkake_flutter/test/date_utils_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:nikkake_flutter/domain/date_utils.dart';
import 'package:nikkake_flutter/models/models.dart';

Routine buildRoutine({
  FrequencyType frequencyType = FrequencyType.daily,
  int frequencyValue = 1,
  List<int> frequencyDays = const [],
  bool isActive = true,
}) {
  final now = DateTime.now().toUtc();
  return Routine(
    id: 'r1',
    name: 'テスト',
    color: '#6C5CE7',
    icon: '💪',
    frequencyType: frequencyType,
    frequencyValue: frequencyValue,
    frequencyDays: frequencyDays,
    isActive: isActive,
    createdAt: now,
    updatedAt: now,
  );
}

RoutineLog buildLog(String logDate) {
  final now = DateTime.now().toUtc();
  return RoutineLog(
    id: 'log-$logDate',
    routineId: 'r1',
    logDate: logDate,
    status: LogStatus.completed,
    createdAt: now,
    updatedAt: now,
  );
}

void main() {
  group('日付ユーティリティ', () {
    test('getDateString はローカルタイムの YYYY-MM-DD を返す', () {
      expect(getDateString(DateTime(2026, 8, 1)), '2026-08-01');
      expect(getDateString(DateTime(2026, 1, 9)), '2026-01-09');
    });

    test('parseDateString はローカルの0時として解釈する', () {
      final parsed = parseDateString('2026-08-01');
      expect(parsed.year, 2026);
      expect(parsed.month, 8);
      expect(parsed.day, 1);
    });

    test('daysBetween は時刻を無視して日数差を返す', () {
      expect(daysBetween(DateTime(2026, 8, 1, 23), DateTime(2026, 8, 2, 1)), 1);
      expect(daysBetween(DateTime(2026, 8, 1), DateTime(2026, 8, 1)), 0);
      expect(daysBetween(DateTime(2026, 8, 5), DateTime(2026, 8, 1)), -4);
    });

    test('addDays は月をまたいでも正しく進む', () {
      expect(getDateString(addDays(DateTime(2026, 7, 30), 3)), '2026-08-02');
    });
  });

  group('isRoutineDueToday', () {
    // 2026-08-01 は土曜日
    final today = DateTime(2026, 8, 1);

    test('停止中のルーティンは対象外', () {
      expect(isRoutineDueToday(buildRoutine(isActive: false), null, today), isFalse);
    });

    test('毎日のルーティンは常に対象', () {
      expect(isRoutineDueToday(buildRoutine(), buildLog('2026-07-31'), today), isTrue);
    });

    test('N日ごとは前回からの間隔で判定する', () {
      final routine = buildRoutine(frequencyType: FrequencyType.everyNDays, frequencyValue: 3);
      expect(isRoutineDueToday(routine, buildLog('2026-07-29'), today), isTrue);
      expect(isRoutineDueToday(routine, buildLog('2026-07-30'), today), isFalse);
    });

    test('N日ごとで一度も実施していなければ今日から始められる', () {
      final routine = buildRoutine(frequencyType: FrequencyType.everyNDays, frequencyValue: 7);
      expect(isRoutineDueToday(routine, null, today), isTrue);
    });

    test('曜日指定は今日の曜日で判定する（日曜=0）', () {
      // 土曜 = 6
      expect(
        isRoutineDueToday(
          buildRoutine(frequencyType: FrequencyType.weekly, frequencyDays: [6]),
          null,
          today,
        ),
        isTrue,
      );
      expect(
        isRoutineDueToday(
          buildRoutine(frequencyType: FrequencyType.weekly, frequencyDays: [1, 3]),
          null,
          today,
        ),
        isFalse,
      );
    });

    test('日曜日が 0 として扱われる', () {
      final sunday = DateTime(2026, 8, 2);
      expect(
        isRoutineDueToday(
          buildRoutine(frequencyType: FrequencyType.weekly, frequencyDays: [0]),
          null,
          sunday,
        ),
        isTrue,
      );
    });

    test('曜日指定で曜日が未設定なら毎日扱い', () {
      expect(
        isRoutineDueToday(buildRoutine(frequencyType: FrequencyType.weekly), null, today),
        isTrue,
      );
    });
  });

  group('表示フォーマット', () {
    test('formatDuration は時間・分・秒を切り替える', () {
      expect(formatDuration(0), '0:00');
      expect(formatDuration(65), '1:05');
      expect(formatDuration(3725), '1:02:05');
      expect(formatDuration(null), '--:--');
    });

    test('formatWeight は不要な小数を落とす', () {
      expect(formatWeight(60), '60 kg');
      expect(formatWeight(62.5), '62.5 kg');
      expect(formatWeight(null), '-- kg');
    });

    test('formatFrequency は頻度を日本語にする', () {
      expect(formatFrequency(buildRoutine()), '毎日');
      expect(
        formatFrequency(buildRoutine(frequencyType: FrequencyType.everyNDays, frequencyValue: 3)),
        '3日ごと',
      );
      expect(
        formatFrequency(buildRoutine(frequencyType: FrequencyType.everyNDays, frequencyValue: 1)),
        '毎日',
      );
      expect(
        formatFrequency(buildRoutine(frequencyType: FrequencyType.weekly, frequencyDays: [3, 1])),
        '月・水',
      );
    });

    test('greetingForHour は時間帯に応じた挨拶を返す', () {
      expect(greetingForHour(3), 'おつかれさま');
      expect(greetingForHour(9), 'おはよう！');
      expect(greetingForHour(14), 'こんにちは！');
      expect(greetingForHour(21), 'こんばんは！');
    });
  });
}

File: nikkake_kmp/shared/src/commonTest/kotlin/com/myapplication/common/DateUtilsTest.kt
package com.myapplication.common

import com.myapplication.common.data.FrequencyType
import com.myapplication.common.domain.addDays
import com.myapplication.common.domain.daysBetween
import com.myapplication.common.domain.formatDuration
import com.myapplication.common.domain.formatFrequency
import com.myapplication.common.domain.formatWeight
import com.myapplication.common.domain.getDateString
import com.myapplication.common.domain.greetingForHour
import com.myapplication.common.domain.isRoutineDueToday
import com.myapplication.common.domain.parseDateString
import kotlinx.datetime.LocalDate
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class DateUtilsTest {

    // 2026-08-01 は土曜日
    private val today = LocalDate(2026, 8, 1)

    @Test
    fun getDateStringは常に2桁ゼロ埋めで返す() {
        assertEquals("2026-08-01", getDateString(LocalDate(2026, 8, 1)))
        assertEquals("2026-01-09", getDateString(LocalDate(2026, 1, 9)))
    }

    @Test
    fun parseDateStringは日付を正しく復元する() {
        val parsed = parseDateString("2026-08-01")
        assertEquals(2026, parsed.year)
        assertEquals(8, parsed.monthNumber)
        assertEquals(1, parsed.dayOfMonth)
    }

    @Test
    fun daysBetweenは日数差を返す() {
        assertEquals(1, daysBetween(LocalDate(2026, 8, 1), LocalDate(2026, 8, 2)))
        assertEquals(0, daysBetween(LocalDate(2026, 8, 1), LocalDate(2026, 8, 1)))
        assertEquals(-4, daysBetween(LocalDate(2026, 8, 5), LocalDate(2026, 8, 1)))
    }

    @Test
    fun addDaysは月をまたいでも正しく進む() {
        assertEquals("2026-08-02", getDateString(addDays(LocalDate(2026, 7, 30), 3)))
    }

    @Test
    fun 停止中のルーティンは今日の対象外() {
        assertFalse(isRoutineDueToday(buildRoutine(isActive = false), null, today))
    }

    @Test
    fun 毎日のルーティンは常に対象() {
        assertTrue(isRoutineDueToday(buildRoutine(), buildLog("2026-07-31"), today))
    }

    @Test
    fun N日ごとは前回からの間隔で判定する() {
        val routine = buildRoutine(frequencyType = FrequencyType.EVERY_N_DAYS, frequencyValue = 3)
        assertTrue(isRoutineDueToday(routine, buildLog("2026-07-29"), today))
        assertFalse(isRoutineDueToday(routine, buildLog("2026-07-30"), today))
    }

    @Test
    fun N日ごとで未実施なら今日から始められる() {
        val routine = buildRoutine(frequencyType = FrequencyType.EVERY_N_DAYS, frequencyValue = 7)
        assertTrue(isRoutineDueToday(routine, null, today))
    }

    @Test
    fun 曜日指定は日曜を0として判定する() {
        // 土曜 = 6
        assertTrue(
            isRoutineDueToday(
                buildRoutine(frequencyType = FrequencyType.WEEKLY, frequencyDays = listOf(6)),
                null,
                today,
            ),
        )
        assertFalse(
            isRoutineDueToday(
                buildRoutine(frequencyType = FrequencyType.WEEKLY, frequencyDays = listOf(1, 3)),
                null,
                today,
            ),
        )

        val sunday = LocalDate(2026, 8, 2)
        assertTrue(
            isRoutineDueToday(
                buildRoutine(frequencyType = FrequencyType.WEEKLY, frequencyDays = listOf(0)),
                null,
                sunday,
            ),
        )
    }

    @Test
    fun 曜日未設定の曜日指定は毎日扱い() {
        assertTrue(isRoutineDueToday(buildRoutine(frequencyType = FrequencyType.WEEKLY), null, today))
    }

    @Test
    fun formatDurationは時間分秒を切り替える() {
        assertEquals("0:00", formatDuration(0))
        assertEquals("1:05", formatDuration(65))
        assertEquals("1:02:05", formatDuration(3725))
        assertEquals("--:--", formatDuration(null))
    }

    @Test
    fun formatWeightは不要な小数を落とす() {
        assertEquals("60 kg", formatWeight(60.0))
        assertEquals("62.5 kg", formatWeight(62.5))
        assertEquals("-- kg", formatWeight(null))
    }

    @Test
    fun formatFrequencyは頻度を日本語にする() {
        assertEquals("毎日", formatFrequency(buildRoutine()))
        assertEquals(
            "3日ごと",
            formatFrequency(buildRoutine(frequencyType = FrequencyType.EVERY_N_DAYS, frequencyValue = 3)),
        )
        assertEquals(
            "毎日",
            formatFrequency(buildRoutine(frequencyType = FrequencyType.EVERY_N_DAYS, frequencyValue = 1)),
        )
        assertEquals(
            "月・水",
            formatFrequency(buildRoutine(frequencyType = FrequencyType.WEEKLY, frequencyDays = listOf(3, 1))),
        )
    }

    @Test
    fun greetingForHourは時間帯に応じた挨拶を返す() {
        assertEquals("おつかれさま", greetingForHour(3))
        assertEquals("おはよう！", greetingForHour(9))
        assertEquals("こんにちは！", greetingForHour(14))
        assertEquals("こんばんは！", greetingForHour(21))
    }
}
